### PR TITLE
Cloak momentum based consumption

### PR DESCRIPTION
**Pull request by Aingar**
By default it replaces the cloak alert sound with cloak entry sound, to restore that, put the following into cloak.cfg

[General]
cloak_alert_sound = dsy_jumpdrive_survey

### CODE REVIEW

File: Plugins/Public/cloak_plugin/Cloak.cpp [hunk 1]
```diff
@@ -100,6 +107,7 @@ static map<uint, CLIENTCDSTRUCT> mapClientsCD;
 static map<uint, CLOAK_ARCH> mapCloakingDevices;
 static map<uint, CDSTRUCT> mapCloakDisruptors;
 
+static uint CloakAlertSound = CreateID("cloak_osiris");
```
Comment (dsyalex) [inline Plugins/Public/cloak_plugin/Cloak.cpp:110]: rename to lowerCamelCase please

File: Plugins/Public/cloak_plugin/Cloak.cpp [hunk 2]
```diff
@@ -291,51 +309,40 @@ void SetState(uint iClientID, uint iShipID, int iNewState)
...
 	if (info.bAdmin)
 		return true;
 
-	if (!info.singleCloakConsumed && info.arch.singleUseCloak) {
-		return removeSingleFuel(iClientID, info);
-	}
-
-	else if (info.singleCloakConsumed)
-	{
+	if(setJumpingClients.find(iClientID) != setJumpingClients.end())
```
Comment (dsyalex) [inline Plugins/Public/cloak_plugin/Cloak.cpp:318]: why are we adding logic around jumping while cloaking here? if you're jumping then you don't need any cloaking fuel removed?

File: Plugins/Public/cloak_plugin/Cloak.cpp [hunk 3]
```diff
@@ -291,51 +309,40 @@ void SetState(uint iClientID, uint iShipID, int iNewState)
...
+			const auto& fuelUsage = info.arch.mapFuelToUsage[item->iArchID];
+			float currFuelUsage = fuelUsage.usageStatic;
+			if (fuelUsage.usageLinear != 0.0f || fuelUsage.usageSquare != 0.0f) {
+				Vector dir1;
+				Vector dir2;
+				pub::SpaceObj::GetMotion(iShipID, dir1, dir2);
+				float vecLength = sqrtf(dir1.x * dir1.x + dir1.y * dir1.y + dir1.z * dir1.z);
+				
+				currFuelUsage += fuelUsage.usageLinear * vecLength;
+				currFuelUsage += fuelUsage.usageSquare * vecLength * vecLength;
```
Comment (dsyalex) [inline Plugins/Public/cloak_plugin/Cloak.cpp:334]: so wait... we want to add fuel usage based on speed... but also based on speed squared?

File: Plugins/Public/cloak_plugin/Cloak.cpp [hunk 4]
```diff
@@ -291,51 +309,40 @@ void SetState(uint iClientID, uint iShipID, int iNewState)
...
+			}
+			uint totalFuelUsage = static_cast<uint>(max(currFuelUsage, 0.0f));
+			if (item->iCount >= totalFuelUsage)
 			{
-				pub::Player::RemoveCargo(iClientID, item->sID, fuel_usage);
+				pub::Player::RemoveCargo(iClientID, item->sID, totalFuelUsage);
 				return true;
 			}
+			if(info.arch.mapFuelToUsage.size() == 1)
```
Comment (dsyalex) [inline Plugins/Public/cloak_plugin/Cloak.cpp:342]: is this some sort of weird optimisation for specific configs?

File: Plugins/Public/cloak_plugin/Cloak.cpp (hunk 2 above)
Comment (Aingar) [reply Plugins/Public/cloak_plugin/Cloak.cpp:318]: Should probably be renamed. When in a jump tunnel, player ship is moving at an insane speed, around 10kps. This was added not to instantly drain their fuel the moment they hit a jumphole/gate

File: Plugins/Public/cloak_plugin/Cloak.cpp (hunk 3 above)
Comment (Aingar) [reply Plugins/Public/cloak_plugin/Cloak.cpp:334]: Correct, we want the formula to scale exponentially with speed, requiring us to use a quadratic formula to achieve wanted consumption at specific speed breakpoints. It also enables the player to traverse a larger distance if they're willing to move at lower speed.

File: Plugins/Public/cloak_plugin/Cloak.cpp (hunk 4 above)
Comment (Aingar) [reply Plugins/Public/cloak_plugin/Cloak.cpp:342]: If we're using a single fuel commodity and we just finished processing it, there's no reason to continue looking, so we break the loop early.

File: Plugins/Public/cloak_plugin/Cloak.cpp (hunk 3 above)
Comment (dsyalex) [reply Plugins/Public/cloak_plugin/Cloak.cpp:334]: very well

File: Plugins/Public/cloak_plugin/Cloak.cpp (hunk 2 above)
Comment (dsyalex) [reply Plugins/Public/cloak_plugin/Cloak.cpp:318]: Oh that's interesting to hear written down and stated as fact. I always had a suspicion that while in the jump tunnel the player is actually still in one of the systems somewhere moving at speed, but never proved it. Makes sense.

File: Plugins/Public/cloak_plugin/Cloak.cpp (hunk 4 above)
Comment (dsyalex) [reply Plugins/Public/cloak_plugin/Cloak.cpp:342]: ok I guess, stops us from having to iterate through the rest of the player's equipDescList

**Review by dsyalex — CHANGES_REQUESTED** [review]
variable renaming and then this is good to go

File: Plugins/Public/cloak_plugin/Cloak.cpp [hunk 5]
```diff
@@ -962,6 +952,14 @@ int __cdecl Dock_Call(unsigned int const &iShip, unsigned int const &iDockTarget
 	return 0;
 }
 
+void __stdcall SystemSwitchOut(uint iClientID, FLPACKET_SYSTEM_SWITCH_OUT& switchOutPacket)
+{
+	// in case of SERVER_PACKET hooks, first argument is junk data before it gets processed by the server.
```
Comment (dsyalex) [inline Plugins/Public/cloak_plugin/Cloak.cpp:957]: what does this mean? iClientID Is wrong and that's why we're going to `HkGetClientIDByShip(switchOutPacket.shipId)` instead??

**Review by dsyalex — CHANGES_REQUESTED** [review]

**Review by dsyalex — APPROVED** [review]

**Pull request merged**